// mind-elixir/dist/MindElixirLite.js
const Wn = 0, Yn = 1, Rn = 2, tt = {
  name: "Latte",
  type: "light",
  palette: ["#dd7878", "#ea76cb", "#8839ef", "#e64553", "#fe640b", "#df8e1d", "#40a02b", "#209fb5", "#1e66f5", "#7287fd"],
  cssVar: {
    "--node-gap-x": "30px",
    "--node-gap-y": "10px",
    "--main-gap-x": "65px",
    "--main-gap-y": "45px",
    "--root-radius": "30px",
    "--main-radius": "20px",
    "--root-color": "#ffffff",
    "--root-bgcolor": "#4c4f69",
    "--root-border-color": "rgba(0, 0, 0, 0)",
    "--main-border": "",
    // you can customize, it will fallback to 2px solid main-color
    "--main-color": "#444446",
    "--main-bgcolor": "#ffffff",
    "--main-bgcolor-transparent": "rgba(255, 255, 255, 0.8)",
    "--topic-padding": "3px",
    "--color": "#777777",
    "--bgcolor": "#f6f6f6",
    "--selected": "#4dc4ff",
    "--accent-color": "#e64553",
    "--panel-color": "#444446",
    "--panel-bgcolor": "#ffffff",
    "--panel-border-color": "#eaeaea",
    "--map-padding": "50px 80px"
  }
}, et = {
  name: "Dark",
  type: "dark",
  palette: ["#848FA0", "#748BE9", "#D2F9FE", "#4145A5", "#789AFA", "#706CF4", "#EF987F", "#775DD5", "#FCEECF", "#DA7FBC"],
  cssVar: {
    "--node-gap-x": "30px",
    "--node-gap-y": "10px",
    "--main-gap-x": "65px",
    "--main-gap-y": "45px",
    "--root-radius": "30px",
    "--main-radius": "20px",
    "--root-color": "#ffffff",
    "--root-bgcolor": "#2d3748",
    "--root-border-color": "rgba(255, 255, 255, 0.1)",
    "--main-border": "",
    "--main-color": "#ffffff",
    "--main-bgcolor": "#4c4f69",
    "--main-bgcolor-transparent": "rgba(76, 79, 105, 0.8)",
    "--topic-padding": "3px",
    "--color": "#cccccc",
    "--bgcolor": "#252526",
    "--selected": "#4dc4ff",
    "--accent-color": "#789AFA",
    "--panel-color": "#ffffff",
    "--panel-bgcolor": "#2d3748",
    "--panel-border-color": "#696969",
    "--map-padding": "50px 80px"
  }
};
function It(t) {
  return t.replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/"/g, "&quot;");
}
const G = function(t, e) {
  if (e.id === t)
    return e;
  if (e.children && e.children.length) {
    for (let n = 0; n < e.children.length; n++) {
      const o = G(t, e.children[n]);
      if (o) return o;
    }
    return null;
  } else
    return null;
}, nt = (t, e) => {
  if (t.parent = e, t.children)
    for (let n = 0; n < t.children.length; n++)
      nt(t.children[n], t);
}, wt = (t, e, n) => {
  t.expanded = e, t.children && t.children.forEach((o) => {
    wt(o, e);
  });
};
function _(t, e, n, o) {
  const s = n - t, i = o - e, c = Math.atan2(i, s) * 180 / Math.PI, l = 12, a = 30, f = (c + 180 - a) * Math.PI / 180, d = (c + 180 + a) * Math.PI / 180;
  return {
    x1: n + Math.cos(f) * l,
    y1: o + Math.sin(f) * l,
    x2: n + Math.cos(d) * l,
    y2: o + Math.sin(d) * l
  };
}
function Y() {
  return ((/* @__PURE__ */ new Date()).getTime().toString(16) + Math.random().toString(16).substring(2)).substring(2, 18);
}
const Ot = function() {
  const t = Y();
  return {
    topic: this.newTopicName,
    id: t
  };
}, P = (t, e) => {
  let n = 0, o = 0;
  for (; e && e !== t; )
    n += e.offsetLeft, o += e.offsetTop, e = e.offsetParent;
  return { offsetLeft: n, offsetTop: o };
}, D = (t, e) => {
  for (const n in e)
    t.setAttribute(n, e[n]);
}, J = (t) => t ? t.tagName === "ME-TPC" : !1, ot = (t) => {
  const e = /translate3d\(([^,]+),\s*([^,]+)/, n = t.match(e);
  return n ? { x: parseFloat(n[1]), y: parseFloat(n[2]) } : { x: 0, y: 0 };
}, bt = function(t) {
  for (let e = 0; e < t.length; e++) {
    const { dom: n, evt: o, func: s } = t[e];
    n.addEventListener(o, s);
  }
  return function() {
    for (let n = 0; n < t.length; n++) {
      const { dom: o, evt: s, func: i } = t[n];
      o.removeEventListener(s, i);
    }
  };
}, ht = (t, e) => {
  const n = t.x - e.x, o = t.y - e.y;
  return Math.sqrt(n * n + o * o);
}, Bt = function(t, e) {
  if (!e)
    return j(t), t;
  let n = t.querySelector(".insert-preview");
  const o = `insert-preview ${e} show`;
  return n || (n = document.createElement("div"), t.appendChild(n)), n.className = o, t;
}, j = function(t) {
  if (!t) return;
  const e = t.querySelectorAll(".insert-preview");
  for (const n of e || [])
    n.remove();
}, dt = function(t, e) {
  for (const n of e) {
    const o = n.parentElement.parentElement.contains(t);
    if (!(t && t.tagName === "ME-TPC" && t !== n && !o && t.nodeObj.parent)) return !1;
  }
  return !0;
}, Wt = function(t) {
  const e = document.createElement("div");
  return e.className = "mind-elixir-ghost", t.container.appendChild(e), e;
};
class Yt {
  mind;
  isMoving = !1;
  interval = null;
  speed = 20;
  constructor(e) {
    this.mind = e;
  }
  move(e, n) {
    this.isMoving || (this.isMoving = !0, this.interval = setInterval(() => {
      this.mind.move(e * this.speed * this.mind.scaleVal, n * this.speed * this.mind.scaleVal);
    }, 100));
  }
  stop() {
    this.isMoving = !1, this.interval && (clearInterval(this.interval), this.interval = null);
  }
}
function Rt(t) {
  return {
    isDragging: !1,
    insertType: null,
    meet: null,
    ghost: Wt(t),
    edgeMoveController: new Yt(t),
    startX: 0,
    startY: 0,
    pointerId: null
  };
}
const Xt = 5;
function ft(t, e, n, o = !1) {
  if (t.spacePressed) return !1;
  const s = n.target;
  if (s?.tagName !== "ME-TPC" || !s.nodeObj.parent) return !1;
  if (e.startX = n.clientX, e.startY = n.clientY, e.pointerId = n.pointerId, t.dragged = t.currentNodes, o) {
    xt(t, e);
    const i = t.container.getBoundingClientRect();
    vt(e.ghost, n.clientX - i.x, n.clientY - i.y);
  }
  return !0;
}
function vt(t, e, n) {
  t.style.transform = `translate(${e - 10}px, ${n - 10}px)`, t.style.display = "block";
}
function xt(t, e) {
  const { dragged: n } = t;
  if (!n) return;
  const o = document.activeElement;
  o && o.isContentEditable && o.blur(), e.isDragging = !0, n.length > 1 ? e.ghost.innerHTML = n.length + "" : e.ghost.innerHTML = n[0].innerHTML;
  for (const s of n)
    s.parentElement.parentElement.style.opacity = "0.5";
  t.panHelper.clear();
}
function Vt(t, e, n) {
  const { dragged: o } = t;
  if (!o || e.pointerId !== n.pointerId) return;
  const s = n.clientX - e.startX, i = n.clientY - e.startY, r = Math.sqrt(s * s + i * i);
  if (!e.isDragging && r > Xt && xt(t, e), !e.isDragging) return;
  const c = t.container.getBoundingClientRect();
  vt(e.ghost, n.clientX - c.x, n.clientY - c.y), n.clientX < c.x + 50 ? e.edgeMoveController.move(1, 0) : n.clientX > c.x + c.width - 50 ? e.edgeMoveController.move(-1, 0) : n.clientY < c.y + 50 ? e.edgeMoveController.move(0, 1) : n.clientY > c.y + c.height - 50 ? e.edgeMoveController.move(0, -1) : e.edgeMoveController.stop(), j(e.meet);
  const l = 12 * t.scaleVal, a = document.elementFromPoint(n.clientX, n.clientY - l);
  if (dt(a, o)) {
    e.meet = a;
    const f = a.getBoundingClientRect(), d = f.y;
    n.clientY > d + f.height ? e.insertType = "after" : e.insertType = "in";
  } else {
    const f = document.elementFromPoint(n.clientX, n.clientY + l);
    if (dt(f, o)) {
      e.meet = f;
      const p = f.getBoundingClientRect().y;
      n.clientY < p ? e.insertType = "before" : e.insertType = "in";
    } else
      e.insertType = null, e.meet = null;
  }
  e.meet && Bt(e.meet, e.insertType);
}
function zt(t, e, n) {
  const { dragged: o } = t;
  if (!(!o || e.pointerId !== n.pointerId)) {
    e.edgeMoveController.stop();
    for (const s of o)
      s.parentElement.parentElement.style.opacity = "1";
    e.ghost.style.display = "none", e.ghost.innerHTML = "", e.isDragging && e.meet && (j(e.meet), e.insertType === "before" ? t.moveNodeBefore(o, e.meet) : e.insertType === "after" ? t.moveNodeAfter(o, e.meet) : e.insertType === "in" && t.moveNodeIn(o, e.meet)), t.dragged = null, e.isDragging = !1, e.insertType = null, e.meet = null, e.pointerId = null;
  }
}
function ut(t, e) {
  const { dragged: n } = t;
  if (n) {
    e.edgeMoveController.stop();
    for (const o of n)
      o.parentElement.parentElement.style.opacity = "1";
    e.meet && j(e.meet), e.ghost.style.display = "none", e.ghost.innerHTML = "", t.dragged = null, e.isDragging = !1, e.insertType = null, e.meet = null, e.pointerId = null;
  }
}
const O = {
  LHS: "lhs",
  RHS: "rhs"
}, Ft = function() {
  this.nodes.innerHTML = "";
  const t = this.createTopic(this.nodeData);
  St.call(this, t, this.nodeData), t.draggable = !1;
  const e = document.createElement("me-root");
  e.appendChild(t);
  const n = this.nodeData.children || [];
  if (this.direction === 2) {
    let o = 0, s = 0;
    n.map((i) => {
      i.direction === 0 ? o += 1 : i.direction === 1 ? s += 1 : o <= s ? (i.direction = 0, o += 1) : (i.direction = 1, s += 1);
    });
  }
  Gt(this, n, e);
}, Gt = function(t, e, n) {
  const o = document.createElement("me-main");
  o.className = O.LHS;
  const s = document.createElement("me-main");
  s.className = O.RHS;
  for (let i = 0; i < e.length; i++) {
    const r = e[i], { grp: c } = t.createWrapper(r);
    t.direction === 2 ? r.direction === 0 ? o.appendChild(c) : s.appendChild(c) : t.direction === 0 ? o.appendChild(c) : s.appendChild(c);
  }
  t.nodes.appendChild(o), t.nodes.appendChild(n), t.nodes.appendChild(s), t.nodes.appendChild(t.lines), t.nodes.appendChild(t.labelContainer);
}, _t = function(t, e) {
  const n = document.createElement("me-children");
  for (let o = 0; o < e.length; o++) {
    const s = e[o], { grp: i } = t.createWrapper(s);
    n.appendChild(i);
  }
  return n;
}, Ct = function(t, e) {
  const o = (this?.el ? this.el : e || document).querySelector(`[data-nodeid="me${t}"]`);
  if (!o) throw new Error(`FindEle: Node ${t} not found, maybe it's collapsed.`);
  return o;
}, St = function(t, e) {
  if (t.innerHTML = "", e.style) {
    const n = e.style;
    for (const o in n)
      t.style[o] = n[o];
  }
  if (e.dangerouslySetInnerHTML) {
    t.innerHTML = e.dangerouslySetInnerHTML;
    return;
  }
  if (e.image) {
    const n = e.image;
    if (n.url && n.width && n.height) {
      const o = document.createElement("img");
      o.src = this.imageProxy ? this.imageProxy(n.url) : n.url, o.style.width = n.width + "px", o.style.height = n.height + "px", n.fit && (o.style.objectFit = n.fit), t.appendChild(o), t.image = o;
    }
  } else t.image && (t.image = void 0);
  {
    const n = document.createElement("span");
    n.className = "text", this.markdown ? n.innerHTML = this.markdown(e.topic, e) : n.textContent = e.topic, t.appendChild(n), t.text = n;
  }
  if (e.hyperLink) {
    const n = document.createElement("a");
    n.className = "hyper-link", n.target = "_blank", n.innerText = "🔗", n.href = e.hyperLink, t.appendChild(n), t.link = n;
  } else t.link && (t.link = void 0);
  if (e.icons && e.icons.length) {
    const n = document.createElement("span");
    n.className = "icons", n.innerHTML = e.icons.map((o) => `<span>${It(o)}</span>`).join(""), t.appendChild(n), t.icons = n;
  } else t.icons && (t.icons = void 0);
  if (e.tags && e.tags.length) {
    const n = document.createElement("div");
    n.className = "tags", e.tags.forEach((o) => {
      const s = document.createElement("span");
      typeof o == "string" ? s.textContent = o : (s.textContent = o.text, o.className && (s.className = o.className), o.style && Object.assign(s.style, o.style)), n.appendChild(s);
    }), t.appendChild(n), t.tags = n;
  } else t.tags && (t.tags = void 0);
}, qt = function(t, e) {
  const n = document.createElement("me-wrapper"), { p: o, tpc: s } = this.createParent(t);
  if (n.appendChild(o), !e && t.children && t.children.length > 0) {
    const i = Zt(t.expanded);
    if (o.appendChild(i), t.expanded !== !1) {
      const r = _t(this, t.children);
      n.appendChild(r);
    }
  }
  return { grp: n, top: o, tpc: s };
}, jt = function(t) {
  const e = document.createElement("me-parent"), n = this.createTopic(t);
  return St.call(this, n, t), e.appendChild(n), { p: e, tpc: n };
}, Ut = function(t) {
  const e = document.createElement("me-children");
  return e.append(...t), e;
}, Kt = function(t) {
  const e = document.createElement("me-tpc");
  return e.nodeObj = t, e.dataset.nodeid = "me" + t.id, e;
};
function Et(t) {
  const e = document.createRange();
  e.selectNodeContents(t);
  const n = window.getSelection();
  n && (n.removeAllRanges(), n.addRange(e));
}
const Jt = function(t) {
  if (!t) return;
  const e = document.createElement("div"), n = t.nodeObj, o = n.topic, { offsetLeft: s, offsetTop: i } = P(this.nodes, t);
  this.nodes.appendChild(e), e.id = "input-box", e.textContent = o, e.contentEditable = "plaintext-only", e.spellcheck = !1;
  const r = getComputedStyle(t);
  e.style.cssText = `
  left: ${s}px;
  top: ${i}px;
  min-width:${t.offsetWidth - 8}px;
  color:${r.color};
  font-size:${r.fontSize};
  padding:${r.padding};
  margin:${r.margin}; 
  background-color:${r.backgroundColor !== "rgba(0, 0, 0, 0)" && r.backgroundColor};
  border: ${r.border};
  border-radius:${r.borderRadius}; `, this.direction === 0 && (e.style.right = "0"), t.style.opacity = "0", Et(e), this.bus.fire("operation", {
    name: "beginEdit",
    obj: t.nodeObj
  }), e.addEventListener("keydown", (c) => {
    c.stopPropagation();
    const l = c.key;
    if (l === "Enter" || l === "Tab") {
      if (c.shiftKey) return;
      c.preventDefault(), e.blur(), this.container.focus();
    } else l === "Escape" && (c.preventDefault(), e.textContent = o, e.blur(), this.container.focus());
  }), e.addEventListener("blur", () => {
    if (!e) return;
    t.style.opacity = "1", e.remove();
    const c = e.innerText?.trim() || "";
    c === o || c === "" || (n.topic = c, this.markdown ? t.text.innerHTML = this.markdown(n.topic, n) : t.text.textContent = c, this.linkDiv(), this.bus.fire("operation", {
      name: "finishEdit",
      obj: n,
      origin: o
    }));
  });
}, Zt = function(t) {
  const e = document.createElement("me-epd");
  return e.expanded = t !== !1, e.className = t !== !1 ? "minus" : "", e;
}, Qt = function(t) {
  const n = t.parentElement.parentElement.lastElementChild;
  n?.tagName === "svg" && n?.remove();
};
function te(t) {
  return {
    nodeData: t.isFocusMode ? t.nodeDataBackup : t.nodeData,
    arrows: t.arrows,
    summaries: t.summaries,
    direction: t.direction,
    theme: t.theme,
    compact: t.compact,
    meta: t.meta
  };
}
const ee = function(t, e = !1) {
  const n = this.container, o = t.getBoundingClientRect(), s = n.getBoundingClientRect();
  if (e || o.top > s.bottom - 50 || o.bottom < s.top + 50 || o.left > s.right - 50 || o.right < s.left + 50) {
    const r = o.left + o.width / 2, c = o.top + o.height / 2, l = s.left + s.width / 2, a = s.top + s.height / 2, f = r - l, d = c - a;
    this.move(-f, -d, !0);
  }
}, ne = function(t, e, n) {
  this.clearSelection(), this.scrollIntoView(t), this.selection?.select(t), e && this.bus.fire("selectNewNode", t.nodeObj);
}, oe = function(t) {
  this.selection?.select(t);
}, se = function(t) {
  this.selection?.deselect(t);
}, ie = function() {
  this.unselectNodes(this.currentNodes), this.unselectSummary(), this.unselectArrow();
}, Tt = function(t) {
  return JSON.stringify(t, (e, n) => {
    if (!(e === "parent" && typeof n != "string"))
      return n;
  });
}, re = function() {
  const t = te(this);
  return Tt(t);
}, ce = function() {
  return JSON.parse(this.getDataString());
}, le = function() {
  this.editable = !0;
}, ae = function() {
  this.editable = !1;
}, he = function(t, e = { x: 0, y: 0 }) {
  if (t < this.scaleMin && t < this.scaleVal || t > this.scaleMax && t > this.scaleVal) return;
  const n = this.container.getBoundingClientRect(), o = e.x ? e.x - n.left - n.width / 2 : 0, s = e.y ? e.y - n.top - n.height / 2 : 0, { dx: i, dy: r } = st(this), c = this.map.style.transform, { x: l, y: a } = ot(c), f = l - i, d = a - r, p = this.scaleVal, w = (-o + f) * (1 - t / p), y = (-s + d) * (1 - t / p);
  this.map.style.transform = `translate3d(${l - w}px, ${a - y}px, 0) scale(${t})`, this.scaleVal = t, this.bus.fire("scale", t);
}, de = function() {
  const t = this.nodes.offsetHeight / this.container.offsetHeight, e = this.nodes.offsetWidth / this.container.offsetWidth, n = 1 / Math.max(1, Math.max(t, e));
  this.scaleVal = n;
  const { dx: o, dy: s } = st(this, !0);
  this.map.style.transform = `translate3d(${o}px, ${s}px, 0) scale(${n})`, this.bus.fire("scale", n);
}, fe = function(t, e, n = !1) {
  const { map: o, scaleVal: s, bus: i, container: r, nodes: c } = this;
  if (n && o.style.transition === "transform 0.3s")
    return;
  const l = o.style.transform;
  let { x: a, y: f } = ot(l);
  const d = r.getBoundingClientRect(), p = c.getBoundingClientRect(), w = p.left < d.right && p.right > d.left, y = p.top < d.bottom && p.bottom > d.top;
  if (w) {
    const u = p.left + t, g = p.right + t;
    (u >= d.right || g <= d.left) && (t = 0);
  }
  if (y) {
    const u = p.top + e, g = p.bottom + e;
    (u >= d.bottom || g <= d.top) && (e = 0);
  }
  a += t, f += e, n && (o.style.transition = "transform 0.3s", setTimeout(() => {
    o.style.transition = "none";
  }, 300)), o.style.transform = `translate3d(${a}px, ${f}px, 0) scale(${s})`, i.fire("move", { dx: t, dy: e });
}, st = (t, e = !1) => {
  const { container: n, map: o, nodes: s } = t;
  let i, r;
  if (t.alignment === "nodes" || e)
    i = (n.offsetWidth - s.offsetWidth) / 2, r = (n.offsetHeight - s.offsetHeight) / 2, o.style.transformOrigin = "50% 50%";
  else {
    const c = o.querySelector("me-root"), l = c.offsetTop, a = c.offsetLeft, f = c.offsetWidth, d = c.offsetHeight;
    i = n.offsetWidth / 2 - a - f / 2, r = n.offsetHeight / 2 - l - d / 2, o.style.transformOrigin = `${a + f / 2}px 50%`;
  }
  return { dx: i, dy: r };
}, ue = function() {
  const { map: t, container: e } = this, { dx: n, dy: o } = st(this);
  e.scrollTop = 0, e.scrollLeft = 0, t.style.transform = `translate3d(${n}px, ${o}px, 0) scale(${this.scaleVal})`;
}, pe = function(t) {
  t(this);
}, ge = function(t) {
  t.nodeObj.parent && (this.clearSelection(), this.tempDirection === null && (this.tempDirection = this.direction), this.isFocusMode || (this.nodeDataBackup = this.nodeData, this.isFocusMode = !0), this.nodeData = t.nodeObj, this.initRight(), this.toCenter());
}, me = function() {
  this.isFocusMode = !1, this.tempDirection !== null && (this.nodeData = this.nodeDataBackup, this.direction = this.tempDirection, this.tempDirection = null, this.refresh(), this.toCenter());
}, ye = function() {
  this.direction = 0, this.refresh(), this.toCenter(), this.bus.fire("changeDirection", this.direction);
}, we = function() {
  this.direction = 1, this.refresh(), this.toCenter(), this.bus.fire("changeDirection", this.direction);
}, be = function() {
  this.direction = 2, this.refresh(), this.toCenter(), this.bus.fire("changeDirection", this.direction);
}, ve = function(t, e) {
  const n = t.nodeObj;
  typeof e == "boolean" ? n.expanded = e : n.expanded !== !1 ? n.expanded = !1 : n.expanded = !0;
  const o = t.getBoundingClientRect(), s = {
    x: o.left,
    y: o.top
  }, i = t.parentNode, r = i.children[1];
  if (r.expanded = n.expanded, r.className = n.expanded ? "minus" : "", Qt(t), n.expanded) {
    const d = this.createChildren(
      n.children.map((p) => this.createWrapper(p).grp)
    );
    i.parentNode.appendChild(d);
  } else
    i.parentNode.children[1].remove();
  this.linkDiv(t.closest("me-main > me-wrapper"));
  const c = t.getBoundingClientRect(), l = {
    x: c.left,
    y: c.top
  }, a = s.x - l.x, f = s.y - l.y;
  this.move(a, f), this.bus.fire("expandNode", n);
}, xe = function(t, e) {
  const n = t.nodeObj, o = t.getBoundingClientRect(), s = {
    x: o.left,
    y: o.top
  };
  wt(n, e ?? !n.expanded), this.refresh();
  const i = this.findEle(n.id).getBoundingClientRect(), r = {
    x: i.left,
    y: i.top
  }, c = s.x - r.x, l = s.y - r.y;
  this.move(c, l);
}, Ce = function(t) {
  this.clearSelection(), t && (t = JSON.parse(JSON.stringify(t)), this.nodeData = t.nodeData, this.arrows = t.arrows || [], this.summaries = t.summaries || [], t.meta && (this.meta = t.meta), t.theme && this.changeTheme(t.theme)), nt(this.nodeData), this.layout(), this.linkDiv();
}, Se = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  cancelFocus: me,
  clearSelection: ie,
  disableEdit: ae,
  enableEdit: le,
  expandNode: ve,
  expandNodeAll: xe,
  focusNode: ge,
  getData: ce,
  getDataString: re,
  initLeft: ye,
  initRight: we,
  initSide: be,
  install: pe,
  move: fe,
  refresh: Ce,
  scale: he,
  scaleFit: de,
  scrollIntoView: ee,
  selectNode: ne,
  selectNodes: oe,
  stringifyData: Tt,
  toCenter: ue,
  unselectNodes: se
}, Symbol.toStringTag, { value: "Module" })), Ee = 40, Te = 10, De = ({ deltaMode: t, deltaY: e, viewportHeight: n }) => t === WheelEvent.DOM_DELTA_LINE ? e * Ee : t === WheelEvent.DOM_DELTA_PAGE ? e * n : e, Le = ({ deltaMode: t, deltaY: e, scaleSensitivity: n, viewportHeight: o }) => {
  const i = -De({ deltaMode: t, deltaY: e, viewportHeight: o }) / Te * n;
  return Math.max(-n, Math.min(n, i));
}, Me = (t, e, n) => {
  e !== 0 && t.scale(t.scaleVal + e, n);
}, Pe = (t, e) => {
  const n = Le({
    deltaMode: e.deltaMode,
    deltaY: e.deltaY,
    scaleSensitivity: t.scaleSensitivity,
    viewportHeight: t.container.clientHeight || window.innerHeight
  });
  Me(t, n, { x: e.clientX, y: e.clientY });
};
function ke(t) {
  const { panHelper: e, container: n } = t;
  let o = null;
  t.spacePressed = !1;
  const s = {
    lastTap: 0,
    lastTapTarget: null,
    DOUBLE_CLICK_THRESHOLD: 300,
    detect(h, m) {
      if (h.button !== 0) {
        this.clear();
        return;
      }
      const x = (/* @__PURE__ */ new Date()).getTime(), E = x - this.lastTap, S = E < this.DOUBLE_CLICK_THRESHOLD && E > 0 && this.lastTapTarget === h.target;
      this.lastTap = x, this.lastTapTarget = h.target, S && m(h);
    },
    clear() {
      this.lastTap = 0, this.lastTapTarget = null;
    }
  }, i = {
    Idle: 0,
    Pinch: 1,
    DragWait: 2,
    Drag: 3,
    Pan: 4,
    BoxSelect: 5
  };
  t.ptState = i.Idle;
  const r = {
    lastDistance: null,
    activePointers: /* @__PURE__ */ new Map(),
    handlePointerDown(h) {
      if (h.pointerType !== "touch") return !1;
      if (this.activePointers.set(h.pointerId, { x: h.clientX, y: h.clientY }), this.activePointers.size >= 2) {
        const [m, x] = Array.from(this.activePointers.values());
        return this.lastDistance = ht(m, x), !0;
      }
      return !1;
    },
    handlePointerMove(h) {
      if (h.pointerType !== "touch" || !this.activePointers.has(h.pointerId)) return !1;
      if (this.activePointers.set(h.pointerId, { x: h.clientX, y: h.clientY }), this.activePointers.size >= 2) {
        const [m, x] = Array.from(this.activePointers.values()), E = ht(m, x);
        if (this.lastDistance !== null && this.lastDistance > 0) {
          const S = E / this.lastDistance;
          t.scale(t.scaleVal * S, {
            x: (m.x + x.x) / 2,
            y: (m.y + x.y) / 2
          });
        }
        return this.lastDistance = E, !0;
      }
      return !1;
    },
    handlePointerUp(h) {
      h.pointerType === "touch" && (this.activePointers.delete(h.pointerId), this.activePointers.size < 2 && (this.lastDistance = null));
    },
    clear() {
      this.activePointers.clear(), this.lastDistance = null;
    }
  }, c = Rt(t), l = {
    timer: null,
    startPos: null,
    pointerId: null,
    DURATION: 500,
    MOVE_THRESHOLD: 10,
    clear() {
      this.timer !== null && (clearTimeout(this.timer), this.timer = null, this.startPos = null, this.pointerId = null);
    },
    start(h, m) {
      this.timer = window.setTimeout(() => {
        m(h), this.timer = null, this.startPos = null, this.pointerId = null;
      }, this.DURATION), this.startPos = { x: h.clientX, y: h.clientY }, this.pointerId = h.pointerId;
    },
    handleMove(h) {
      if (this.timer !== null && this.startPos !== null && h.pointerId === this.pointerId) {
        const m = h.clientX - this.startPos.x, x = h.clientY - this.startPos.y;
        Math.sqrt(m * m + x * x) > this.MOVE_THRESHOLD && this.clear();
      }
    }
  }, a = (h, m) => {
    if (h.closest("#input-box")) return !1;
    const x = h.closest(".svg-label"), E = h.closest(".topiclinks, .summary"), S = x ? { type: x.dataset.type, element: document.getElementById(x.dataset.svgId) } : E ? { type: E.classList.contains("topiclinks") ? "arrow" : "summary", element: h.closest("g") } : null;
    if (!S?.type || !S?.element) return !1;
    const { type: M, element: N } = S;
    return t.clearSelection(), M === "arrow" ? m ? t.editArrowLabel(N) : t.selectArrow(N) : m ? t.editSummary(N) : t.selectSummary(N), !0;
  }, f = (h) => {
    if (h.pointerType === "mouse" && h.button !== 0) return;
    if (t.helper1?.moved) {
      t.helper1.clear();
      return;
    }
    if (t.helper2?.moved) {
      t.helper2.clear();
      return;
    }
    if (e.moved) {
      e.clear();
      return;
    }
    if (c?.isDragging)
      return;
    const m = h.target;
    m.tagName === "ME-EPD" && (h.ctrlKey || h.metaKey ? t.expandNodeAll(m.previousSibling) : t.expandNode(m.previousSibling));
  }, d = (h) => {
    if (!t.editable) return;
    const m = h.target;
    if (J(m)) {
      t.selectNode(m), t.beginEdit(m);
      return;
    }
    a(m, !0);
  }, p = (h) => {
    if (h.pointerType === "touch" && r.handlePointerDown(h)) {
      t.ptState = i.Pinch, l.clear(), e.clear(), (c.isDragging || c.pointerId !== null) && ut(t, c);
      return;
    }
    if (t.ptState === i.Pinch) return;
    const m = h.target;
    if (t.editable && m.className === "map-container" && h.button === 0 && h.pointerType === "mouse") {
      t.ptState = i.BoxSelect;
      return;
    }
    if (e.handlePointerDown(h), e.mousedown && (t.ptState = i.Pan), h.button === 0 || h.pointerType === "touch")
      if (J(m)) {
        t.selection?.cancel();
        const E = t.currentNodes || [];
        if (h.ctrlKey || h.metaKey || t.mobileMultiSelect ? E.includes(m) ? o = m : ((t.currentArrow || t.currentSummary) && t.clearSelection(), t.selection?.select(m)) : E.includes(m) || t.selectNode(m), !t.editable) return;
        h.pointerType === "touch" ? (t.ptState = i.DragWait, l.start(h, (M) => {
          ft(t, c, M, !0) && (t.ptState = i.Drag, m.setPointerCapture(M.pointerId));
        })) : ft(t, c, h, !1) && (t.ptState = i.Drag, m.setPointerCapture(h.pointerId));
      } else
        a(m, !1);
  }, w = (h) => {
    switch (t.ptState) {
      case i.Pinch:
        r.handlePointerMove(h);
        break;
      case i.DragWait:
        l.handleMove(h), l.timer === null && (t.ptState = i.Idle);
        break;
      case i.Drag:
        Vt(t, c, h);
        break;
      case i.Pan:
        e.handlePointerMove(h);
        break;
    }
  }, y = (h) => {
    h.pointerType === "touch" && r.handlePointerUp(h);
    const m = c.isDragging, x = e.moved;
    switch (t.ptState) {
      case i.DragWait:
        l.clear();
        break;
      case i.Drag:
        zt(t, c, h);
        break;
      case i.Pan:
        e.handlePointerUp(h);
        break;
    }
    s.detect(h, d), (t.ptState !== i.Pinch || r.activePointers.size < 2) && (t.ptState = i.Idle), o && (!m && !x && t.selection?.deselect(o), o = null);
  }, u = () => {
    r.clear(), l.clear(), e.clear(), s.clear(), (c.isDragging || c.pointerId !== null) && ut(t, c), t.ptState = i.Idle, o = null;
  }, g = (h) => {
    h.preventDefault(), h.button === 2 && t.editable && setTimeout(() => {
      if (t.panHelper.moved || t.ptState !== i.Idle && t.ptState !== i.Pan) return;
      const m = h.target;
      J(m) && !m.classList.contains("selected") && t.selectNode(m), t.bus.fire("showContextMenu", h);
    }, 200);
  }, v = (h) => {
    if (h.stopPropagation(), h.preventDefault(), h.ctrlKey || h.metaKey) return Pe(t, h);
    if (h.shiftKey) return t.move(-h.deltaY, 0);
    t.move(-h.deltaX, -h.deltaY);
  }, b = (h) => {
    h.code === "Space" && (t.spacePressed = !0, t.container.classList.add("space-pressed"));
  }, C = (h) => {
    h.code === "Space" && (t.spacePressed = !1, t.container.classList.remove("space-pressed"));
  };
  return bt([
    { dom: n, evt: "pointerdown", func: p },
    { dom: n, evt: "pointermove", func: w },
    { dom: n, evt: "pointerup", func: y },
    { dom: n, evt: "pointercancel", func: u },
    { dom: n, evt: "click", func: f },
    { dom: n, evt: "contextmenu", func: g },
    { dom: n, evt: "wheel", func: typeof t.handleWheel == "function" ? t.handleWheel : v },
    { dom: n, evt: "blur", func: u },
    { dom: n, evt: "keydown", func: b },
    { dom: n, evt: "keyup", func: C }
  ]);
}
function Ne() {
  return {
    handlers: {},
    addListener: function(t, e) {
      this.handlers[t] === void 0 && (this.handlers[t] = []), this.handlers[t].push(e);
    },
    fire: function(t, ...e) {
      if (this.handlers[t] instanceof Array) {
        const n = this.handlers[t];
        for (let o = 0; o < n.length; o++)
          n[o](...e);
      }
    },
    removeListener: function(t, e) {
      if (!this.handlers[t]) return;
      const n = this.handlers[t];
      if (!e)
        n.length = 0;
      else if (n.length)
        for (let o = 0; o < n.length; o++)
          n[o] === e && this.handlers[t].splice(o, 1);
    }
  };
}
const k = "http://www.w3.org/2000/svg", U = function(t) {
  const e = t.clientWidth, n = t.clientHeight, o = t.dataset, s = Number(o.x), i = Number(o.y), r = o.anchor;
  let c = s;
  r === "middle" ? c = s - e / 2 : r === "end" && (c = s - e), t.style.left = `${c}px`, t.style.top = `${i - n / 2}px`, t.style.visibility = "visible";
}, Q = function(t, e, n, o) {
  const { anchor: s = "middle", color: i, dataType: r, svgId: c } = o, l = document.createElement("div");
  l.className = "svg-label", l.style.color = i || "#666";
  const a = "label-" + c;
  return l.id = a, l.innerHTML = t, l.dataset.type = r, l.dataset.svgId = c, l.dataset.x = e.toString(), l.dataset.y = n.toString(), l.dataset.anchor = s, l;
}, Dt = function(t, e, n) {
  const o = document.createElementNS(k, "path");
  return D(o, {
    d: t,
    stroke: e || "#666",
    fill: "none",
    "stroke-width": n
  }), o;
}, W = function(t) {
  const e = document.createElementNS(k, "svg");
  return e.setAttribute("class", t), e.setAttribute("overflow", "visible"), e;
}, pt = function() {
  const t = document.createElementNS(k, "line");
  return t.setAttribute("stroke", "#4dc4ff"), t.setAttribute("fill", "none"), t.setAttribute("stroke-width", "2"), t.setAttribute("opacity", "0.45"), t;
}, Ae = function(t, e, n, o) {
  const s = document.createElementNS(k, "g");
  return [
    {
      name: "line",
      d: t
    },
    {
      name: "arrow1",
      d: e
    },
    {
      name: "arrow2",
      d: n
    }
  ].forEach((r, c) => {
    const l = r.d, a = document.createElementNS(k, "path"), f = {
      d: l,
      stroke: o?.stroke || "rgb(227, 125, 116)",
      fill: "none",
      "stroke-linecap": o?.strokeLinecap || "cap",
      "stroke-width": String(o?.strokeWidth || "2")
    };
    o?.opacity !== void 0 && (f.opacity = String(o.opacity)), D(a, f), c === 0 && a.setAttribute("stroke-dasharray", o?.strokeDasharray || "8,2");
    const d = document.createElementNS(k, "path");
    D(d, {
      d: l,
      stroke: "transparent",
      fill: "none",
      "stroke-width": "15"
    }), s.appendChild(d), s.appendChild(a), s[r.name] = a;
  }), s;
}, Lt = function(t, e, n) {
  if (!e) return;
  const o = n.label;
  e.style.opacity = "0";
  const s = e.cloneNode(!0);
  t.nodes.appendChild(s), s.id = "input-box", s.textContent = o, s.contentEditable = "plaintext-only", s.spellcheck = !1, s.style.cssText = `
    left:${e.style.left};
    top:${e.style.top}; 
    max-width: 200px;
  `, Et(s), t.scrollIntoView(s), s.addEventListener("keydown", (i) => {
    i.stopPropagation();
    const r = i.key;
    if (r === "Enter" || r === "Tab") {
      if (i.shiftKey) return;
      i.preventDefault(), s.blur(), t.container.focus();
    }
  }), s.addEventListener("blur", () => {
    if (!s) return;
    const i = s.innerText?.trim() || "";
    i === "" ? n.label = o : n.label = i, e.style.opacity = "1", s.remove(), i !== o && (t.markdown ? e.innerHTML = t.markdown(n.label, n) : e.textContent = n.label, U(e), "parent" in n ? t.bus.fire("operation", {
      name: "finishEditSummary",
      obj: n
    }) : t.bus.fire("operation", {
      name: "finishEditArrowLabel",
      obj: n
    }));
  });
}, $e = function(t) {
  const e = this.map.querySelector("me-root"), n = e.offsetTop, o = e.offsetLeft, s = e.offsetWidth, i = e.offsetHeight, r = this.map.querySelectorAll("me-main > me-wrapper");
  this.lines.innerHTML = "";
  for (let c = 0; c < r.length; c++) {
    const l = r[c], a = l.querySelector("me-tpc"), { offsetLeft: f, offsetTop: d } = P(this.nodes, a), p = a.offsetWidth, w = a.offsetHeight, y = l.parentNode.className, u = this.generateMainBranch({ pT: n, pL: o, pW: s, pH: i, cT: d, cL: f, cW: p, cH: w, direction: y, containerHeight: this.nodes.offsetHeight }), g = this.theme.palette, v = a.nodeObj.branchColor || g[c % g.length];
    if (a.style.borderColor = v, this.lines.appendChild(Dt(u, v, "3")), t && t !== l)
      continue;
    const b = W("subLines"), C = l.lastChild;
    C.tagName === "svg" && C.remove(), l.appendChild(b), Mt(this, b, v, l, y, !0);
  }
  this.labelContainer.innerHTML = "", this.renderArrow(), this.renderSummary(), this.bus.fire("linkDiv");
}, Mt = function(t, e, n, o, s, i) {
  const r = o.firstChild, c = o.children[1].children;
  if (c.length === 0) return;
  const l = r.offsetTop, a = r.offsetLeft, f = r.offsetWidth, d = r.offsetHeight;
  for (let p = 0; p < c.length; p++) {
    const w = c[p], y = w.firstChild, u = y.offsetTop, g = y.offsetLeft, v = y.offsetWidth, b = y.offsetHeight, C = y.firstChild.nodeObj.branchColor || n, T = t.generateSubBranch({ pT: l, pL: a, pW: f, pH: d, cT: u, cL: g, cW: v, cH: b, direction: s, isFirst: i });
    e.appendChild(Dt(T, C, "2"));
    const h = y.children[1];
    if (h) {
      if (!h.expanded) continue;
    } else
      continue;
    Mt(t, e, C, w, s);
  }
}, He = '<?xml version="1.0" standalone="no"?><!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd"><svg t="1750169394918" class="icon" viewBox="0 0 1024 1024" version="1.1" xmlns="http://www.w3.org/2000/svg" p-id="2021" xmlns:xlink="http://www.w3.org/1999/xlink" width="200" height="200"><path d="M851.91168 328.45312c-59.97056 0-108.6208 48.47104-108.91264 108.36992l-137.92768 38.4a109.14304 109.14304 0 0 0-63.46752-46.58688l1.39264-137.11872c47.29344-11.86816 82.31936-54.66624 82.31936-105.64096 0-60.15488-48.76288-108.91776-108.91776-108.91776s-108.91776 48.76288-108.91776 108.91776c0 49.18784 32.60928 90.75712 77.38368 104.27392l-1.41312 138.87488a109.19936 109.19936 0 0 0-63.50336 48.55808l-138.93632-39.48544 0.01024-0.72704c0-60.15488-48.76288-108.91776-108.91776-108.91776s-108.91776 48.75776-108.91776 108.91776c0 60.15488 48.76288 108.91264 108.91776 108.91264 39.3984 0 73.91232-20.92032 93.03552-52.2496l139.19232 39.552-0.00512 0.2304c0 25.8304 9.00096 49.5616 24.02816 68.23424l-90.14272 132.63872a108.7488 108.7488 0 0 0-34.2528-5.504c-60.15488 0-108.91776 48.768-108.91776 108.91776 0 60.16 48.76288 108.91776 108.91776 108.91776 60.16 0 108.92288-48.75776 108.92288-108.91776 0-27.14624-9.9328-51.968-26.36288-71.04l89.04704-131.03104a108.544 108.544 0 0 0 37.6832 6.70208 108.672 108.672 0 0 0 36.48512-6.272l93.13792 132.57216a108.48256 108.48256 0 0 0-24.69888 69.0688c0 60.16 48.768 108.92288 108.91776 108.92288 60.16 0 108.91776-48.76288 108.91776-108.92288 0-60.14976-48.75776-108.91776-108.91776-108.91776a108.80512 108.80512 0 0 0-36.69504 6.3488l-93.07136-132.48a108.48768 108.48768 0 0 0 24.79616-72.22784l136.09984-37.888c18.99008 31.93856 53.84192 53.3504 93.69088 53.3504 60.16 0 108.92288-48.75776 108.92288-108.91264-0.00512-60.15488-48.77312-108.92288-108.92288-108.92288z" p-id="2022"></path></svg>', Ie = '<?xml version="1.0" standalone="no"?><!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd"><svg t="1750169375313" class="icon" viewBox="0 0 1024 1024" version="1.1" xmlns="http://www.w3.org/2000/svg" p-id="1775" xmlns:xlink="http://www.w3.org/1999/xlink" width="200" height="200"><path d="M639 463.30000001L639 285.1c0-36.90000001-26.4-68.5-61.3-68.5l-150.2 0c-1.5 0-3 0.1-4.5 0.3-10.2-38.7-45.5-67.3-87.5-67.3-50 0-90.5 40.5-90.5 90.5s40.5 90.5 90.5 90.5c42 0 77.3-28.6 87.5-67.39999999 1.4 0.3 2.9 0.4 4.5 0.39999999L577.7 263.6c6.8 0 14.3 8.9 14.3 21.49999999l0 427.00000001c0 12.7-7.40000001 21.5-14.30000001 21.5l-150.19999999 0c-1.5 0-3 0.2-4.5 0.4-10.2-38.8-45.5-67.3-87.5-67.3-50 0-90.5 40.5-90.5 90.4 0 49.9 40.5 90.6 90.5 90.59999999 42 0 77.3-28.6 87.5-67.39999999 1.4 0.2 2.9 0.4 4.49999999 0.4L577.7 780.7c34.80000001 0 61.3-31.6 61.3-68.50000001L639 510.3l79.1 0c10.4 38.5 45.49999999 67 87.4 67 50 0 90.5-40.5 90.5-90.5s-40.5-90.5-90.5-90.5c-41.79999999 0-77.00000001 28.4-87.4 67L639 463.30000001z" fill="currentColor" p-id="1776"></path></svg>', Oe = '<?xml version="1.0" standalone="no"?><!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd"><svg t="1750169667709" class="icon" viewBox="0 0 1024 1024" version="1.1" xmlns="http://www.w3.org/2000/svg" p-id="3037" xmlns:xlink="http://www.w3.org/1999/xlink" width="200" height="200"><path d="M385 560.69999999L385 738.9c0 36.90000001 26.4 68.5 61.3 68.5l150.2 0c1.5 0 3-0.1 4.5-0.3 10.2 38.7 45.5 67.3 87.5 67.3 50 0 90.5-40.5 90.5-90.5s-40.5-90.5-90.5-90.5c-42 0-77.3 28.6-87.5 67.39999999-1.4-0.3-2.9-0.4-4.5-0.39999999L446.3 760.4c-6.8 0-14.3-8.9-14.3-21.49999999l0-427.00000001c0-12.7 7.40000001-21.5 14.30000001-21.5l150.19999999 0c1.5 0 3-0.2 4.5-0.4 10.2 38.8 45.5 67.3 87.5 67.3 50 0 90.5-40.5 90.5-90.4 0-49.9-40.5-90.6-90.5-90.59999999-42 0-77.3 28.6-87.5 67.39999999-1.4-0.2-2.9-0.4-4.49999999-0.4L446.3 243.3c-34.80000001 0-61.3 31.6-61.3 68.50000001L385 513.7l-79.1 0c-10.4-38.5-45.49999999-67-87.4-67-50 0-90.5 40.5-90.5 90.5s40.5 90.5 90.5 90.5c41.79999999 0 77.00000001-28.4 87.4-67L385 560.69999999z" fill="currentColor" p-id="3038"></path></svg>', Be = '<?xml version="1.0" standalone="no"?><!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd"><svg t="1750169402629" class="icon" viewBox="0 0 1024 1024" version="1.1" xmlns="http://www.w3.org/2000/svg" p-id="2170" xmlns:xlink="http://www.w3.org/1999/xlink" width="200" height="200"><path d="M639.328 416c8.032 0 16.096-3.008 22.304-9.056l202.624-197.184-0.8 143.808c-0.096 17.696 14.144 32.096 31.808 32.192 0.064 0 0.128 0 0.192 0 17.6 0 31.904-14.208 32-31.808l1.248-222.208c0-0.672-0.352-1.248-0.384-1.92 0.032-0.512 0.288-0.896 0.288-1.408 0.032-17.664-14.272-32-31.968-32.032L671.552 96l-0.032 0c-17.664 0-31.968 14.304-32 31.968C639.488 145.632 653.824 160 671.488 160l151.872 0.224-206.368 200.8c-12.672 12.32-12.928 32.608-0.64 45.248C622.656 412.736 630.976 416 639.328 416z" p-id="2171"></path><path d="M896.032 639.552 896.032 639.552c-17.696 0-32 14.304-32.032 31.968l-0.224 151.872-200.832-206.4c-12.32-12.64-32.576-12.96-45.248-0.64-12.672 12.352-12.928 32.608-0.64 45.248l197.184 202.624-143.808-0.8c-0.064 0-0.128 0-0.192 0-17.6 0-31.904 14.208-32 31.808-0.096 17.696 14.144 32.096 31.808 32.192l222.24 1.248c0.064 0 0.128 0 0.192 0 0.64 0 1.12-0.32 1.76-0.352 0.512 0.032 0.896 0.288 1.408 0.288l0.032 0c17.664 0 31.968-14.304 32-31.968L928 671.584C928.032 653.952 913.728 639.584 896.032 639.552z" p-id="2172"></path><path d="M209.76 159.744l143.808 0.8c0.064 0 0.128 0 0.192 0 17.6 0 31.904-14.208 32-31.808 0.096-17.696-14.144-32.096-31.808-32.192L131.68 95.328c-0.064 0-0.128 0-0.192 0-0.672 0-1.248 0.352-1.888 0.384-0.448 0-0.8-0.256-1.248-0.256 0 0-0.032 0-0.032 0-17.664 0-31.968 14.304-32 31.968L96 352.448c-0.032 17.664 14.272 32 31.968 32.032 0 0 0.032 0 0.032 0 17.664 0 31.968-14.304 32-31.968l0.224-151.936 200.832 206.4c6.272 6.464 14.624 9.696 22.944 9.696 8.032 0 16.096-3.008 22.304-9.056 12.672-12.32 12.96-32.608 0.64-45.248L209.76 159.744z" p-id="2173"></path><path d="M362.368 617.056l-202.624 197.184 0.8-143.808c0.096-17.696-14.144-32.096-31.808-32.192-0.064 0-0.128 0-0.192 0-17.6 0-31.904 14.208-32 31.808l-1.248 222.24c0 0.704 0.352 1.312 0.384 2.016 0 0.448-0.256 0.832-0.256 1.312-0.032 17.664 14.272 32 31.968 32.032L352.448 928c0 0 0.032 0 0.032 0 17.664 0 31.968-14.304 32-31.968s-14.272-32-31.968-32.032l-151.936-0.224 206.4-200.832c12.672-12.352 12.96-32.608 0.64-45.248S375.008 604.704 362.368 617.056z" p-id="2174"></path></svg>', We = '<?xml version="1.0" standalone="no"?><!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd"><svg t="1750169573443" class="icon" viewBox="0 0 1024 1024" version="1.1" xmlns="http://www.w3.org/2000/svg" p-id="2883" xmlns:xlink="http://www.w3.org/1999/xlink" width="200" height="200"><path d="M514.133333 488.533333m-106.666666 0a106.666667 106.666667 0 1 0 213.333333 0 106.666667 106.666667 0 1 0-213.333333 0Z" fill="currentColor" p-id="2884"></path><path d="M512 64C264.533333 64 64 264.533333 64 512c0 236.8 183.466667 428.8 416 445.866667v-134.4c-53.333333-59.733333-200.533333-230.4-200.533333-334.933334 0-130.133333 104.533333-234.666667 234.666666-234.666666s234.666667 104.533333 234.666667 234.666666c0 61.866667-49.066667 153.6-145.066667 270.933334l-59.733333 68.266666V960C776.533333 942.933333 960 748.8 960 512c0-247.466667-200.533333-448-448-448z" fill="currentColor" p-id="2885"></path></svg>', Ye = '<?xml version="1.0" standalone="no"?><!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd"><svg t="1750169419447" class="icon" viewBox="0 0 1024 1024" version="1.1" xmlns="http://www.w3.org/2000/svg" p-id="2480" xmlns:xlink="http://www.w3.org/1999/xlink" width="200" height="200"><path d="M863.328 482.56l-317.344-1.12L545.984 162.816c0-17.664-14.336-32-32-32s-32 14.336-32 32l0 318.4L159.616 480.064c-0.032 0-0.064 0-0.096 0-17.632 0-31.936 14.24-32 31.904C127.424 529.632 141.728 544 159.392 544.064l322.592 1.152 0 319.168c0 17.696 14.336 32 32 32s32-14.304 32-32l0-318.944 317.088 1.12c0.064 0 0.096 0 0.128 0 17.632 0 31.936-14.24 32-31.904C895.264 496.992 880.96 482.624 863.328 482.56z" p-id="2481"></path></svg>', Re = '<?xml version="1.0" standalone="no"?><!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd"><svg t="1750169426515" class="icon" viewBox="0 0 1024 1024" version="1.1" xmlns="http://www.w3.org/2000/svg" p-id="2730" xmlns:xlink="http://www.w3.org/1999/xlink" width="200" height="200"><path d="M863.744 544 163.424 544c-17.664 0-32-14.336-32-32s14.336-32 32-32l700.32 0c17.696 0 32 14.336 32 32S881.44 544 863.744 544z" p-id="2731"></path></svg>', Xe = {
  side: He,
  left: Ie,
  right: Oe,
  full: Be,
  living: We,
  zoomin: Ye,
  zoomout: Re
}, H = (t, e) => {
  const n = document.createElement("span");
  return n.id = t, n.innerHTML = Xe[e], n;
};
function Ve(t) {
  const e = document.createElement("div"), n = H("fullscreen", "full"), o = H("toCenter", "living"), s = H("zoomout", "zoomout"), i = H("zoomin", "zoomin");
  e.appendChild(n), e.appendChild(o), e.appendChild(s), e.appendChild(i), e.className = "mind-elixir-toolbar rb";
  let r = null;
  const c = () => {
    const a = t.container.getBoundingClientRect(), f = ot(t.map.style.transform), d = a.width / 2, p = a.height / 2, w = (d - f.x) / t.scaleVal, y = (p - f.y) / t.scaleVal;
    r = {
      containerRect: a,
      currentTransform: f,
      mapCenterX: w,
      mapCenterY: y
    };
  }, l = () => {
    if (r) {
      const a = t.container.getBoundingClientRect(), f = a.width / 2, d = a.height / 2, p = f - r.mapCenterX * t.scaleVal, w = d - r.mapCenterY * t.scaleVal, y = p - r.currentTransform.x, u = w - r.currentTransform.y;
      t.move(y, u);
    }
  };
  return t.el.addEventListener("fullscreenchange", l), n.onclick = () => {
    c(), document.fullscreenElement !== t.el ? t.el.requestFullscreen() : document.exitFullscreen();
  }, o.onclick = () => {
    t.toCenter();
  }, s.onclick = () => {
    t.scale(t.scaleVal - t.scaleSensitivity);
  }, i.onclick = () => {
    t.scale(t.scaleVal + t.scaleSensitivity);
  }, e;
}
function ze(t) {
  const e = document.createElement("div"), n = H("tbltl", "left"), o = H("tbltr", "right"), s = H("tblts", "side");
  return e.appendChild(n), e.appendChild(o), e.appendChild(s), e.className = "mind-elixir-toolbar lt", n.onclick = () => {
    t.initLeft();
  }, o.onclick = () => {
    t.initRight();
  }, s.onclick = () => {
    t.initSide();
  }, e;
}
function Fe(t) {
  t.container.append(Ve(t)), t.container.append(ze(t));
}
function Pt({ pT: t, pL: e, pW: n, pH: o, cT: s, cL: i, cW: r, cH: c, direction: l, containerHeight: a }) {
  let f = e + n / 2;
  const d = t + o / 2;
  let p;
  l === O.LHS ? p = i + r : p = i;
  const w = s + c / 2, u = (1 - Math.abs(w - d) / a) * 0.25 * (n / 2);
  return l === O.LHS ? f = f - n / 10 - u : f = f + n / 10 + u, `M ${f} ${d} Q ${f} ${w} ${p} ${w}`;
}
function kt({ pT: t, pL: e, pW: n, pH: o, cT: s, cL: i, cW: r, cH: c, direction: l, isFirst: a }) {
  const f = parseInt(this.container.style.getPropertyValue("--node-gap-x"));
  let d = 0, p = 0;
  a ? d = t + o / 2 : d = t + o;
  const w = s + c;
  let y = 0, u = 0, g = 0;
  const v = Math.abs(d - w) / 300 * f;
  return l === O.LHS ? (g = e, y = g + f, u = g - f, p = i + f, `M ${y} ${d} C ${g} ${d} ${g + v} ${w} ${u} ${w} H ${p}`) : (g = e + n, y = g - f, u = g + f, p = i + r - f, `M ${y} ${d} C ${g} ${d} ${g - v} ${w} ${u} ${w} H ${p}`);
}
const Ge = function(t, e = !0) {
  this.theme = t, this.generateMainBranch = this.theme.generateMainBranch || Pt, this.generateSubBranch = this.theme.generateSubBranch || kt;
  const o = {
    ...(this.theme.type === "dark" ? et : tt).cssVar,
    ...this.theme.cssVar
  };
  this.compact && (o["--node-gap-x"] = "15px", o["--node-gap-y"] = "2px", o["--main-gap-x"] = "30px", o["--main-gap-y"] = "6px");
  const s = Object.keys(o);
  for (let i = 0; i < s.length; i++) {
    const r = s[i];
    this.container.style.setProperty(r, o[r]);
  }
  e && this.refresh();
}, _e = function(t) {
  this.compact = t, this.theme && this.changeTheme(this.theme);
}, qe = function(t) {
  return {
    dom: t,
    moved: !1,
    // differentiate click and move
    pointerdown: !1,
    lastX: 0,
    lastY: 0,
    handlePointerMove(e) {
      if (this.pointerdown) {
        this.moved = !0;
        const n = e.clientX - this.lastX, o = e.clientY - this.lastY;
        this.lastX = e.clientX, this.lastY = e.clientY, this.cb && this.cb(n, o);
      }
    },
    handlePointerDown(e) {
      e.button === 0 && (this.pointerdown = !0, this.lastX = e.clientX, this.lastY = e.clientY, this.dom.setPointerCapture(e.pointerId));
    },
    handleClear(e) {
      this.pointerdown = !1, e.pointerId !== void 0 && this.dom.releasePointerCapture(e.pointerId);
    },
    cb: null,
    init(e, n) {
      this.cb = n, this.handleClear = this.handleClear.bind(this), this.handlePointerMove = this.handlePointerMove.bind(this), this.handlePointerDown = this.handlePointerDown.bind(this), this.destroy = bt([
        { dom: e, evt: "pointermove", func: this.handlePointerMove },
        { dom: e, evt: "pointerleave", func: this.handleClear },
        { dom: e, evt: "pointerup", func: this.handleClear },
        { dom: this.dom, evt: "pointerdown", func: this.handlePointerDown }
      ]);
    },
    destroy: null,
    clear() {
      this.moved = !1, this.pointerdown = !1;
    }
  };
}, gt = {
  create: qe
}, Nt = "#4dc4ff";
function At(t, e, n, o, s, i, r, c) {
  return {
    x: t / 8 + n * 3 / 8 + s * 3 / 8 + r / 8,
    y: e / 8 + o * 3 / 8 + i * 3 / 8 + c / 8
  };
}
function je(t, e, n) {
  t && (t.dataset.x = e.toString(), t.dataset.y = n.toString(), U(t));
}
function V(t, e, n, o, s) {
  D(t, {
    x1: e + "",
    y1: n + "",
    x2: o + "",
    y2: s + ""
  });
}
function mt(t, e, n, o, s, i, r, c, l, a) {
  const f = `M ${e} ${n} C ${o} ${s} ${i} ${r} ${c} ${l}`;
  if (t.line.setAttribute("d", f), a.style) {
    const u = a.style;
    u.stroke && t.line.setAttribute("stroke", u.stroke), u.strokeWidth && t.line.setAttribute("stroke-width", String(u.strokeWidth)), u.strokeDasharray && t.line.setAttribute("stroke-dasharray", u.strokeDasharray), u.strokeLinecap && t.line.setAttribute("stroke-linecap", u.strokeLinecap), u.opacity !== void 0 && t.line.setAttribute("opacity", String(u.opacity));
  }
  const d = t.querySelectorAll('path[stroke="transparent"]');
  d.length > 0 && d[0].setAttribute("d", f);
  const p = _(i, r, c, l);
  if (p) {
    const u = `M ${p.x1} ${p.y1} L ${c} ${l} L ${p.x2} ${p.y2}`;
    if (t.arrow1.setAttribute("d", u), d.length > 1 && d[1].setAttribute("d", u), a.style) {
      const g = a.style;
      g.stroke && t.arrow1.setAttribute("stroke", g.stroke), g.strokeWidth && t.arrow1.setAttribute("stroke-width", String(g.strokeWidth)), g.strokeLinecap && t.arrow1.setAttribute("stroke-linecap", g.strokeLinecap), g.opacity !== void 0 && t.arrow1.setAttribute("opacity", String(g.opacity));
    }
  }
  if (a.bidirectional) {
    const u = _(o, s, e, n);
    if (u) {
      const g = `M ${u.x1} ${u.y1} L ${e} ${n} L ${u.x2} ${u.y2}`;
      if (t.arrow2.setAttribute("d", g), d.length > 2 && d[2].setAttribute("d", g), a.style) {
        const v = a.style;
        v.stroke && t.arrow2.setAttribute("stroke", v.stroke), v.strokeWidth && t.arrow2.setAttribute("stroke-width", String(v.strokeWidth)), v.strokeLinecap && t.arrow2.setAttribute("stroke-linecap", v.strokeLinecap), v.opacity !== void 0 && t.arrow2.setAttribute("opacity", String(v.opacity));
      }
    }
  }
  const { x: w, y } = At(e, n, o, s, i, r, c, l);
  if (t.labelEl && je(t.labelEl, w, y), a.style?.labelColor) {
    const u = t.labelEl;
    u && (u.style.color = a.style.labelColor);
  }
  nn(t);
}
function q(t, e, n) {
  const { offsetLeft: o, offsetTop: s } = P(t.nodes, e), i = e.offsetWidth, r = e.offsetHeight, c = o + i / 2, l = s + r / 2, a = c + n.x, f = l + n.y;
  return {
    w: i,
    h: r,
    cx: c,
    cy: l,
    ctrlX: a,
    ctrlY: f
  };
}
function I(t) {
  let e, n;
  const o = (t.cy - t.ctrlY) / (t.ctrlX - t.cx);
  return o > t.h / t.w || o < -t.h / t.w ? t.cy - t.ctrlY < 0 ? (e = t.cx - t.h / 2 / o, n = t.cy + t.h / 2) : (e = t.cx + t.h / 2 / o, n = t.cy - t.h / 2) : t.cx - t.ctrlX < 0 ? (e = t.cx + t.w / 2, n = t.cy - t.w * o / 2) : (e = t.cx - t.w / 2, n = t.cy + t.w * o / 2), {
    x: e,
    y: n
  };
}
const Ue = function(t, e, n) {
  const o = P(t.nodes, e), s = P(t.nodes, n), i = o.offsetLeft + e.offsetWidth / 2, r = o.offsetTop + e.offsetHeight / 2, c = s.offsetLeft + n.offsetWidth / 2, l = s.offsetTop + n.offsetHeight / 2, a = c - i, f = l - r, d = Math.sqrt(a * a + f * f), p = Math.max(50, Math.min(200, d * 0.3)), w = Math.abs(a), y = Math.abs(f);
  let u, g;
  if (d < 150) {
    const b = e.closest("me-main").className === "lhs" ? -1 : 1;
    u = { x: 200 * b, y: 0 }, g = { x: 200 * b, y: 0 };
  } else if (w > y * 1.5) {
    const b = a > 0 ? e.offsetWidth / 2 : -e.offsetWidth / 2, C = a > 0 ? -n.offsetWidth / 2 : n.offsetWidth / 2;
    u = { x: b + (a > 0 ? p : -p), y: 0 }, g = { x: C + (a > 0 ? -p : p), y: 0 };
  } else if (y > w * 1.5) {
    const b = f > 0 ? e.offsetHeight / 2 : -e.offsetHeight / 2, C = f > 0 ? -n.offsetHeight / 2 : n.offsetHeight / 2;
    u = { x: 0, y: b + (f > 0 ? p : -p) }, g = { x: 0, y: C + (f > 0 ? -p : p) };
  } else {
    const b = Math.atan2(f, a), C = e.offsetWidth / 2 * Math.cos(b), T = e.offsetHeight / 2 * Math.sin(b), h = -(n.offsetWidth / 2) * Math.cos(b), m = -(n.offsetHeight / 2) * Math.sin(b), x = p * 0.7 * (a > 0 ? 1 : -1), E = p * 0.7 * (f > 0 ? 1 : -1);
    u = { x: C + x, y: T + E }, g = { x: h - x, y: m - E };
  }
  return {
    delta1: { x: Math.round(u.x), y: Math.round(u.y) },
    delta2: { x: Math.round(g.x), y: Math.round(g.y) }
  };
}, it = function(t, e, n, o, s) {
  if (!e || !n)
    return;
  if (!o.delta1 || !o.delta2) {
    const S = Ue(t, e, n);
    o.delta1 = S.delta1, o.delta2 = S.delta2;
  }
  const i = q(t, e, o.delta1), r = q(t, n, o.delta2), { x: c, y: l } = I(i), { ctrlX: a, ctrlY: f } = i, { ctrlX: d, ctrlY: p } = r, { x: w, y } = I(r), u = _(d, p, w, y);
  if (!u) return;
  const g = `M ${u.x1} ${u.y1} L ${w} ${y} L ${u.x2} ${u.y2}`;
  let v = "";
  if (o.bidirectional) {
    const S = _(a, f, c, l);
    if (!S) return;
    v = `M ${S.x1} ${S.y1} L ${c} ${l} L ${S.x2} ${S.y2}`;
  }
  const b = Ae(`M ${c} ${l} C ${a} ${f} ${d} ${p} ${w} ${y}`, g, v, o.style), { x: C, y: T } = At(c, l, a, f, d, p, w, y), h = o.style?.labelColor || "rgb(235, 95, 82)", m = "a-" + o.id;
  b.id = m;
  const x = t.markdown ? t.markdown(o.label, o) : o.label, E = Q(x, C, T, {
    anchor: "middle",
    color: h,
    dataType: "arrow",
    svgId: m
  });
  b.labelEl = E, b.arrowObj = o, b.dataset.linkid = o.id, t.labelContainer.appendChild(E), t.arrowSvg.appendChild(b), U(E), s || (t.arrows.push(o), t.currentArrow = b, Ht(t, o, i, r));
}, Ke = function(t, e, n = {}) {
  const o = {
    id: Y(),
    label: "Custom Link",
    from: t.nodeObj.id,
    to: e.nodeObj.id,
    ...n
  };
  it(this, t, e, o), this.bus.fire("operation", {
    name: "createArrow",
    obj: o
  });
}, Je = function(t) {
  K(this);
  const e = { ...t, id: Y() };
  it(this, this.findEle(e.from), this.findEle(e.to), e), this.bus.fire("operation", {
    name: "createArrow",
    obj: e
  });
}, Ze = function(t) {
  let e;
  if (t ? e = t : e = this.currentArrow, !e) return;
  K(this);
  const n = e.arrowObj.id;
  this.arrows = this.arrows.filter((o) => o.id !== n), e.labelEl?.remove(), e.remove(), this.bus.fire("operation", {
    name: "removeArrow",
    obj: {
      id: n
    }
  });
}, Qe = function(t) {
  this.currentArrow = t;
  const e = t.arrowObj, n = this.findEle(e.from), o = this.findEle(e.to), s = q(this, n, e.delta1), i = q(this, o, e.delta2);
  this.editable ? Ht(this, e, s, i) : $t(t, Nt), this.bus.fire("selectArrow", e);
}, tn = function() {
  K(this), this.currentArrow = null, this.bus.fire("unselectArrow");
}, Z = function(t, e) {
  const n = document.createElementNS(k, "path");
  return D(n, {
    d: t,
    stroke: e,
    fill: "none",
    "stroke-width": "6",
    "stroke-linecap": "round",
    "stroke-linejoin": "round"
  }), n;
}, $t = function(t, e) {
  const n = document.createElementNS(k, "g");
  n.setAttribute("class", "arrow-highlight"), n.setAttribute("opacity", "0.45");
  const o = Z(t.line.getAttribute("d"), e);
  n.appendChild(o);
  const s = Z(t.arrow1.getAttribute("d"), e);
  if (n.appendChild(s), t.arrow2.getAttribute("d")) {
    const i = Z(t.arrow2.getAttribute("d"), e);
    n.appendChild(i);
  }
  t.insertBefore(n, t.firstChild);
}, en = function(t) {
  const e = t.querySelector(".arrow-highlight");
  e && e.remove();
}, nn = function(t) {
  const e = t.querySelector(".arrow-highlight");
  if (!e) return;
  const n = e.querySelectorAll("path");
  n.length >= 1 && n[0].setAttribute("d", t.line.getAttribute("d")), n.length >= 2 && n[1].setAttribute("d", t.arrow1.getAttribute("d")), n.length >= 3 && t.arrow2.getAttribute("d") && n[2].setAttribute("d", t.arrow2.getAttribute("d"));
}, K = function(t) {
  t.helper1?.destroy(), t.helper2?.destroy(), t.linkController.style.display = "none", t.P2.style.display = "none", t.P3.style.display = "none", t.currentArrow && en(t.currentArrow);
}, Ht = function(t, e, n, o) {
  const { linkController: s, P2: i, P3: r, line1: c, line2: l, nodes: a, map: f, currentArrow: d, bus: p } = t;
  if (!d) return;
  s.style.display = "initial", i.style.display = "initial", r.style.display = "initial", a.appendChild(s), a.appendChild(i), a.appendChild(r), $t(d, Nt);
  let { x: w, y } = I(n), { ctrlX: u, ctrlY: g } = n, { ctrlX: v, ctrlY: b } = o, { x: C, y: T } = I(o);
  i.style.cssText = `top:${g}px;left:${u}px;`, r.style.cssText = `top:${b}px;left:${v}px;`, V(c, w, y, u, g), V(l, v, b, C, T), t.helper1 = gt.create(i), t.helper2 = gt.create(r), t.helper1.init(f, (h, m) => {
    u = u + h / t.scaleVal, g = g + m / t.scaleVal;
    const x = I({ ...n, ctrlX: u, ctrlY: g });
    w = x.x, y = x.y, i.style.top = g + "px", i.style.left = u + "px", mt(d, w, y, u, g, v, b, C, T, e), V(c, w, y, u, g), e.delta1.x = Math.round(u - n.cx), e.delta1.y = Math.round(g - n.cy), p.fire("updateArrowDelta", e);
  }), t.helper2.init(f, (h, m) => {
    v = v + h / t.scaleVal, b = b + m / t.scaleVal;
    const x = I({ ...o, ctrlX: v, ctrlY: b });
    C = x.x, T = x.y, r.style.top = b + "px", r.style.left = v + "px", mt(d, w, y, u, g, v, b, C, T, e), V(l, v, b, C, T), e.delta2.x = Math.round(v - o.cx), e.delta2.y = Math.round(b - o.cy), p.fire("updateArrowDelta", e);
  });
};
function on() {
  this.arrowSvg.innerHTML = "", this.labelContainer.querySelectorAll('.svg-label[data-type="arrow"]').forEach((e) => e.remove());
  for (let e = 0; e < this.arrows.length; e++) {
    const n = this.arrows[e];
    try {
      it(this, this.findEle(n.from), this.findEle(n.to), n, !0);
    } catch {
    }
  }
  this.nodes.appendChild(this.arrowSvg);
}
function sn(t) {
  K(this), t && t.labelEl && Lt(this, t.labelEl, t.arrowObj);
}
function rn() {
  this.arrows = this.arrows.filter((t) => G(t.from, this.nodeData) && G(t.to, this.nodeData));
}
const cn = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  createArrow: Ke,
  createArrowFrom: Je,
  editArrowLabel: sn,
  removeArrow: Ze,
  renderArrow: on,
  selectArrow: Qe,
  tidyArrow: rn,
  unselectArrow: tn
}, Symbol.toStringTag, { value: "Module" })), ln = function(t) {
  if (t.length === 0) throw new Error("No selected node.");
  if (t.length === 1) {
    const l = t[0].nodeObj, a = t[0].nodeObj.parent;
    if (!a) throw new Error("Can not select root node.");
    const f = a.children.findIndex((d) => l === d);
    return {
      parent: a.id,
      start: f,
      end: f
    };
  }
  let e = 0;
  const n = t.map((l) => {
    let a = l.nodeObj;
    const f = [];
    for (; a.parent; ) {
      const d = a.parent, w = d.children?.indexOf(a);
      a = d, f.unshift({ node: a, index: w });
    }
    return f.length > e && (e = f.length), f;
  });
  let o = 0;
  t: for (; o < e; o++) {
    const l = n[0][o]?.node;
    for (let a = 1; a < n.length; a++)
      if (n[a][o]?.node !== l)
        break t;
  }
  if (!o) throw new Error("Can not select root node.");
  const s = n.map((l) => l[o - 1].index).sort(), i = s[0] || 0, r = s[s.length - 1] || 0, c = n[0][o - 1].node;
  if (!c.parent) throw new Error("Please select nodes in the same main topic.");
  return {
    parent: c.id,
    start: i,
    end: r
  };
}, an = function(t) {
  const e = document.createElementNS(k, "g");
  return e.setAttribute("id", t), e;
}, yt = function(t, e) {
  const n = document.createElementNS(k, "path");
  return D(n, {
    d: t,
    stroke: e || "#666",
    fill: "none",
    "stroke-linecap": "round",
    "stroke-width": "2"
  }), n;
}, hn = (t) => t.parentElement.parentElement, dn = function(t, { parent: e, start: n }) {
  const o = t.findEle(e), s = o.nodeObj;
  let i;
  return s.parent ? i = o.closest("me-main").className : i = t.findEle(s.children[n].id).closest("me-main").className, i;
}, rt = function(t, e) {
  const { id: n, label: o, parent: s, start: i, end: r, style: c } = e, { nodes: l, theme: a, summarySvg: f } = t, p = t.findEle(s).nodeObj, w = dn(t, e);
  let y = 1 / 0, u = 0, g = 0, v = 0;
  for (let B = i; B <= r; B++) {
    const ct = p.children?.[B];
    if (!ct)
      return t.removeSummary(n), null;
    const R = hn(t.findEle(ct.id)), { offsetLeft: X, offsetTop: lt } = P(l, R), at = i === r ? 10 : 20;
    B === i && (g = lt + at), B === r && (v = lt + R.offsetHeight - at), X < y && (y = X), R.offsetWidth + X > u && (u = R.offsetWidth + X);
  }
  let b, C;
  const T = p.parent ? 10 : 0, h = g + T, m = v + T, x = (h + m) / 2, E = c?.stroke || a.cssVar["--color"], S = c?.labelColor || a.cssVar["--color"], M = "s-" + n, N = t.markdown ? t.markdown(o, e) : o;
  w === O.LHS ? (b = yt(`M ${y + 10} ${h} c -5 0 -10 5 -10 10 L ${y} ${m - 10} c 0 5 5 10 10 10 M ${y} ${x} h -10`, E), C = Q(N, y - 20, x, { anchor: "end", color: S, dataType: "summary", svgId: M })) : (b = yt(`M ${u - 10} ${h} c 5 0 10 5 10 10 L ${u} ${m - 10} c 0 5 -5 10 -10 10 M ${u} ${x} h 10`, E), C = Q(N, u + 20, x, { anchor: "start", color: S, dataType: "summary", svgId: M }));
  const $ = an(M);
  return $.appendChild(b), t.labelContainer.appendChild(C), U(C), $.summaryObj = e, $.labelEl = C, f.appendChild($), $;
}, fn = function(t = {}) {
  if (!this.currentNodes) return;
  const { currentNodes: e, summaries: n, bus: o } = this, { parent: s, start: i, end: r } = ln(e), c = { id: Y(), parent: s, start: i, end: r, label: "summary", style: t.style }, l = rt(this, c);
  n.push(c), this.editSummary(l), o.fire("operation", {
    name: "createSummary",
    obj: c
  });
}, un = function(t) {
  const e = Y(), n = { ...t, id: e };
  rt(this, n), this.summaries.push(n), this.bus.fire("operation", {
    name: "createSummary",
    obj: n
  });
}, pn = function(t) {
  const e = this.summaries.findIndex((n) => n.id === t);
  e > -1 && (this.summaries.splice(e, 1), this.nodes.querySelector("#s-" + t)?.remove(), this.nodes.querySelector("#label-s-" + t)?.remove()), this.bus.fire("operation", {
    name: "removeSummary",
    obj: { id: t }
  });
}, gn = function(t) {
  const e = t.labelEl;
  e && e.classList.add("selected"), this.currentSummary = t, this.bus.fire("selectSummary", t.summaryObj);
}, mn = function() {
  this.currentSummary?.labelEl?.classList.remove("selected"), this.currentSummary = null, this.bus.fire("unselectSummary");
}, yn = function() {
  this.summarySvg.innerHTML = "", this.summaries.forEach((t) => {
    try {
      rt(this, t);
    } catch {
    }
  }), this.nodes.insertAdjacentElement("beforeend", this.summarySvg);
}, wn = function(t) {
  t && t.labelEl && Lt(this, t.labelEl, t.summaryObj);
}, bn = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  createSummary: fn,
  createSummaryFrom: un,
  editSummary: wn,
  removeSummary: pn,
  renderSummary: yn,
  selectSummary: gn,
  unselectSummary: mn
}, Symbol.toStringTag, { value: "Module" })), L = "http://www.w3.org/2000/svg";
function vn(t, e) {
  const n = document.createElementNS(L, "svg");
  return D(n, {
    version: "1.1",
    xmlns: L,
    height: t,
    width: e
  }), n;
}
function xn(t, e) {
  return (parseInt(t) - parseInt(e)) / 2;
}
function Cn(t, e, n, o) {
  const s = document.createElementNS(L, "g");
  let i = "";
  return t.text ? i = t.text.textContent : i = t.childNodes[0].textContent, i.split(`
`).forEach((c, l) => {
    const a = document.createElementNS(L, "text");
    D(a, {
      x: n + parseInt(e.paddingLeft) + "",
      y: o + parseInt(e.paddingTop) + xn(e.lineHeight, e.fontSize) * (l + 1) + parseFloat(e.fontSize) * (l + 1) + "",
      "text-anchor": "start",
      "font-family": e.fontFamily,
      "font-size": `${e.fontSize}`,
      "font-weight": `${e.fontWeight}`,
      fill: `${e.color}`
    }), a.innerHTML = c, s.appendChild(a);
  }), s;
}
function Sn(t, e, n, o) {
  let s = "";
  t.nodeObj?.dangerouslySetInnerHTML ? s = t.nodeObj.dangerouslySetInnerHTML : t.text ? s = t.text.textContent : s = t.childNodes[0].textContent;
  const i = document.createElementNS(L, "foreignObject");
  D(i, {
    x: n + parseInt(e.paddingLeft) + "",
    y: o + parseInt(e.paddingTop) + "",
    width: e.width,
    height: e.height
  });
  const r = document.createElement("div");
  return D(r, {
    xmlns: "http://www.w3.org/1999/xhtml",
    style: `font-family: ${e.fontFamily}; font-size: ${e.fontSize}; font-weight: ${e.fontWeight}; color: ${e.color}; white-space: pre-wrap;`
  }), r.innerHTML = s, i.appendChild(r), i;
}
function En(t, e) {
  const n = getComputedStyle(e), { offsetLeft: o, offsetTop: s } = P(t.nodes, e), i = document.createElementNS(L, "rect");
  return D(i, {
    x: o + "",
    y: s + "",
    rx: n.borderRadius,
    ry: n.borderRadius,
    width: n.width,
    height: n.height,
    fill: n.backgroundColor,
    stroke: n.borderColor,
    "stroke-width": n.borderWidth
  }), i;
}
function z(t, e, n = !1) {
  const o = getComputedStyle(e), { offsetLeft: s, offsetTop: i } = P(t.nodes, e), r = document.createElementNS(L, "rect");
  D(r, {
    x: s + "",
    y: i + "",
    rx: o.borderRadius,
    ry: o.borderRadius,
    width: o.width,
    height: o.height,
    fill: o.backgroundColor,
    stroke: o.borderColor,
    "stroke-width": o.borderWidth
  });
  const c = document.createElementNS(L, "g");
  c.appendChild(r);
  let l;
  return n ? l = Sn(e, o, s, i) : l = Cn(e, o, s, i), c.appendChild(l), c;
}
function Tn(t, e) {
  const n = getComputedStyle(e), { offsetLeft: o, offsetTop: s } = P(t.nodes, e), i = document.createElementNS(L, "a"), r = document.createElementNS(L, "text");
  return D(r, {
    x: o + "",
    y: s + parseInt(n.fontSize) + "",
    "text-anchor": "start",
    "font-family": n.fontFamily,
    "font-size": `${n.fontSize}`,
    "font-weight": `${n.fontWeight}`,
    fill: `${n.color}`
  }), r.innerHTML = e.textContent, i.appendChild(r), i.setAttribute("href", e.href), i;
}
function Dn(t, e) {
  const n = getComputedStyle(e), { offsetLeft: o, offsetTop: s } = P(t.nodes, e), i = document.createElementNS(L, "image");
  return D(i, {
    x: o + "",
    y: s + "",
    width: n.width + "",
    height: n.height + "",
    href: e.src
  }), i;
}
const F = 100, Ln = '<?xml version="1.0" standalone="no"?><!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd">', Mn = (t, e = !1) => {
  const n = t.nodes, o = n.offsetHeight + F * 2, s = n.offsetWidth + F * 2, i = vn(o + "px", s + "px"), r = document.createElementNS(L, "svg"), c = document.createElementNS(L, "rect");
  D(c, {
    x: "0",
    y: "0",
    width: `${s}`,
    height: `${o}`,
    fill: t.theme.cssVar["--bgcolor"]
  }), i.appendChild(c), n.querySelectorAll(".subLines").forEach((d) => {
    const p = d.cloneNode(!0), { offsetLeft: w, offsetTop: y } = P(n, d.parentElement);
    p.setAttribute("x", `${w}`), p.setAttribute("y", `${y}`), r.appendChild(p);
  });
  const l = n.querySelector(".lines")?.cloneNode(!0);
  l && r.appendChild(l);
  const a = n.querySelector(".topiclinks")?.cloneNode(!0);
  a && r.appendChild(a);
  const f = n.querySelector(".summary")?.cloneNode(!0);
  return f && r.appendChild(f), n.querySelectorAll("me-tpc").forEach((d) => {
    d.nodeObj.dangerouslySetInnerHTML ? r.appendChild(z(t, d, !e)) : (r.appendChild(En(t, d)), r.appendChild(z(t, d.text, !e)));
  }), n.querySelectorAll(".tags > span").forEach((d) => {
    r.appendChild(z(t, d));
  }), n.querySelectorAll(".icons > span").forEach((d) => {
    r.appendChild(z(t, d));
  }), n.querySelectorAll(".hyper-link").forEach((d) => {
    r.appendChild(Tn(t, d));
  }), n.querySelectorAll("img").forEach((d) => {
    r.appendChild(Dn(t, d));
  }), D(r, {
    x: F + "",
    y: F + "",
    overflow: "visible"
  }), i.appendChild(r), i;
}, Pn = (t, e) => (e && t.insertAdjacentHTML("afterbegin", "<style>" + e + "</style>"), Ln + t.outerHTML);
function kn(t) {
  return new Promise((e, n) => {
    const o = new FileReader();
    o.onload = (s) => {
      e(s.target.result);
    }, o.onerror = (s) => {
      n(s);
    }, o.readAsDataURL(t);
  });
}
const Nn = function(t = !1, e) {
  const n = Mn(this, t), o = Pn(n, e);
  return new Blob([o], { type: "image/svg+xml" });
}, An = async function(t = !1, e) {
  const n = this.exportSvg(t, e), o = await kn(n);
  return new Promise((s, i) => {
    const r = new Image();
    r.setAttribute("crossOrigin", "anonymous"), r.onload = () => {
      const c = document.createElement("canvas");
      c.width = r.width, c.height = r.height, c.getContext("2d").drawImage(r, 0, 0), c.toBlob(s, "image/png", 1);
    }, r.src = o, r.onerror = i;
  });
}, $n = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  exportPng: An,
  exportSvg: Nn
}, Symbol.toStringTag, { value: "Module" })), Hn = {}, In = {
  getObjById: G,
  generateNewObj: Ot,
  layout: Ft,
  linkDiv: $e,
  editTopic: Jt,
  createWrapper: qt,
  createParent: jt,
  createChildren: Ut,
  createTopic: Kt,
  findEle: Ct,
  changeTheme: Ge,
  changeCompact: _e,
  ...Se,
  ...Hn,
  ...cn,
  ...bn,
  ...$n,
  init(t) {
    if (t = JSON.parse(JSON.stringify(t)), !t || !t.nodeData) return new Error("MindElixir: `data` is required");
    t.direction !== void 0 && (this.direction = t.direction), t.compact !== void 0 && (this.compact = t.compact), this.changeTheme(t.theme || this.theme, !1), t.meta && (this.meta = t.meta), this.nodeData = t.nodeData, nt(this.nodeData), this.arrows = t.arrows || [], this.summaries = t.summaries || [], this.tidyArrow(), this.toolBar && Fe(this), this.layout(), this.linkDiv(), this.toCenter();
  },
  destroy() {
    this.disposable.forEach((t) => t()), this.el && (this.el.innerHTML = ""), this.el = void 0, this.nodeData = void 0, this.arrows = void 0, this.summaries = void 0, this.currentArrow = void 0, this.currentNodes = void 0, this.currentSummary = void 0, this.theme = void 0, this.direction = void 0, this.bus = void 0, this.container = void 0, this.map = void 0, this.lines = void 0, this.linkController = void 0, this.arrowSvg = void 0, this.P2 = void 0, this.P3 = void 0, this.line1 = void 0, this.line2 = void 0, this.nodes = void 0, this.selection?.destroy(), this.selection = void 0;
  },
  /**
   * @public
   * @param {boolean} enable
   */
  enableMobileMultiSelect(t) {
    this.mobileMultiSelect = t;
  }
}, On = "5.12.2";
function Bn(t) {
  return {
    x: 0,
    y: 0,
    moved: !1,
    // differentiate click and move
    mousedown: !1,
    handlePointerDown(e) {
      this.moved = !1;
      const n = e.target, o = t.mouseSelectionButton === 0 ? 2 : 0, s = t.spacePressed && e.button === 0 && e.pointerType === "mouse", i = !t.editable || e.button === o && e.pointerType === "mouse" || e.pointerType === "touch";
      !s && !i || (this.x = e.clientX, this.y = e.clientY, n.className !== "circle" && n.contentEditable !== "plaintext-only" && (this.mousedown = !0, n.setPointerCapture(e.pointerId)));
    },
    handlePointerMove(e) {
      if (!this.mousedown || e.target.contentEditable === "plaintext-only" && !t.spacePressed) return !1;
      const n = e.clientX - this.x, o = e.clientY - this.y;
      return this.x = e.clientX, this.y = e.clientY, this.moved = !0, t.move(n, o), !0;
    },
    handlePointerUp(e) {
      if (!this.mousedown) return;
      const n = e.target;
      n.hasPointerCapture && n.hasPointerCapture(e.pointerId) && n.releasePointerCapture(e.pointerId), this.mousedown = !1;
    },
    clear() {
      this.mousedown = !1, this.moved = !1;
    }
  };
}
function A({
  el: t,
  direction: e,
  editable: n,
  contextMenu: o,
  toolBar: s,
  keypress: i,
  mouseSelectionButton: r,
  selectionContainer: c,
  before: l,
  newTopicName: a,
  allowUndo: f,
  generateMainBranch: d,
  generateSubBranch: p,
  overflowHidden: w,
  compact: y,
  theme: u,
  alignment: g,
  scaleSensitivity: v,
  scaleMax: b,
  scaleMin: C,
  handleWheel: T,
  markdown: h,
  imageProxy: m,
  pasteHandler: x,
  mobileMultiSelect: E
}) {
  let S = null;
  const M = Object.prototype.toString.call(t);
  if (M === "[object HTMLDivElement]" ? S = t : M === "[object String]" && (S = document.querySelector(t)), !S) throw new Error("MindElixir: el is not a valid element");
  S.style.position = "relative", S.innerHTML = "", this.el = S, this.disposable = [], this.before = l || {}, this.newTopicName = a || "New Node", this.contextMenu = o ?? !0, this.toolBar = s ?? !0, this.keypress = i ?? !0, this.mouseSelectionButton = r ?? 0, this.direction = e ?? 1, this.editable = n ?? !0, this.allowUndo = f ?? !0, this.scaleSensitivity = v ?? 0.1, this.scaleMax = b ?? 1.4, this.scaleMin = C ?? 0.2, this.generateMainBranch = d || Pt, this.generateSubBranch = p || kt, this.overflowHidden = w ?? !1, this.compact = y ?? !1, this.alignment = g ?? "root", this.handleWheel = T ?? !0, this.markdown = h || void 0, this.imageProxy = m || void 0, this.currentNodes = [], this.currentArrow = null, this.scaleVal = 1, this.tempDirection = null, this.mobileMultiSelect = E ?? !1, this.panHelper = Bn(this), this.bus = Ne(), this.container = document.createElement("div"), this.selectionContainer = c || this.container, this.container.className = "map-container";
  const N = window.matchMedia("(prefers-color-scheme: dark)");
  this.theme = u || (N.matches ? et : tt);
  const $ = document.createElement("div");
  $.className = "map-canvas", this.map = $, this.container.setAttribute("tabindex", "0"), this.container.appendChild(this.map), this.el.appendChild(this.container), this.nodes = document.createElement("me-nodes"), this.lines = W("lines"), this.summarySvg = W("summary"), this.linkController = W("linkcontroller"), this.P2 = document.createElement("div"), this.P3 = document.createElement("div"), this.P2.className = this.P3.className = "circle", this.P2.style.display = this.P3.style.display = "none", this.line1 = pt(), this.line2 = pt(), this.linkController.appendChild(this.line1), this.linkController.appendChild(this.line2), this.arrowSvg = W("topiclinks"), this.labelContainer = document.createElement("div"), this.labelContainer.className = "label-container", this.map.appendChild(this.nodes), this.overflowHidden ? this.container.style.overflow = "hidden" : this.disposable.push(ke(this)), x && (this.pasteHandler = x);
}
A.prototype = In;
Object.defineProperty(A.prototype, "currentNode", {
  get() {
    return this.currentNodes[this.currentNodes.length - 1];
  },
  enumerable: !0
});
A.LEFT = 0;
A.RIGHT = 1;
A.SIDE = 2;
A.THEME = tt;
A.DARK_THEME = et;
A.version = On;
A.E = Ct;
export {
  et as DARK_THEME,
  Wn as LEFT,
  Yn as RIGHT,
  Rn as SIDE,
  tt as THEME,
  A as default
};
